xarray/core/dataset.py

```diff
@@ -581,14 +589,26 @@ def __init__(
         data_vars: Mapping[Any, Any] | None = None,
         coords: Mapping[Any, Any] | None = None,
         attrs: Mapping[Any, Any] | None = None,
+        indexes: Mapping[Any, Index] | None = None,
     ) -> None:
-        # TODO(shoyer): expose indexes as a public argument in __init__
-
         if data_vars is None:
             data_vars = {}
         if coords is None:
             coords = {}
 
+        if indexes is None:
+            create_default_indexes = True
+            indexes = Indexes()
+        elif len(indexes) == 0:
+            create_default_indexes = False
+            indexes = Indexes()
+        else:
+            create_default_indexes = True
+            if not isinstance(indexes, Indexes):
+                raise TypeError("non-empty indexes must be an instance of `Indexes`")
+            elif indexes._index_type != Index:
+                raise TypeError("indexes must only contain Xarray `Index` objects")
```

Comment:
I don't like the special case for size 0 indexes. These sort of special cases that violate type stability can lead to very hard to debug bugs.

Instead, how about never creating default pandas indexes if `indexes` is provided? Maybe there could also be a special method (e.g., `assign_default_indexes()`) for manually adding in pandas indexes later.

Reply:
Yes I like that!

I also like `assign_default_indexes()` or maybe `assign_indexes(indexes=None)` if we don't want to add another method.

Reply:
One concern I have is when a user wants to explicitly provide a pandas multi-index or a custom single coordinate index for a dimension and still wants the default pandas indexes for the other dimension coordinates.

In this case, calling `assign_default_indexes()` or `assign_indexes(indexes=None)` later will likely overwrite the explicitly provided indexes?

After removing the pandas multi-index dimension coordinate this won't be an issue anymore, but the issue remains for any custom 1-d dimension coordinate index.

Should we add an `overwrite=True` keyword argument to `assign_(default_)indexes`?

Reply:
`assign_default_indexes()` should probably include an optional list of dimension names for which to assign default indexes.

By default, we might make it only add indexes for dimensions that doesn't already have an index.